### PR TITLE
Radio Group

### PR DESCRIPTION
**Pull request by felipowsky**
Hi.
First, thank you so much for this control.
I would like to suggest a new way to structure it since I think the current way has some downsides when it comes to selecting radio buttons programmatically.
Currently developers link the buttons in `otherButtons` and the selection is handled by the `DLRadioButton`.
My suggestion is to move this handling to a new class called `DLRadioGroup`.
This way all the responsibility of selecting is managed by `DLRadioGroup` and the `DLRadioButton` holds only the component's appearance.
`DLRadioButton` requires a `DLRadioGroup` for that reason.

You can still configure everything by Interface Builder using the Object component to represent the `DLRadioGroup`.
pr1
pr2

Here is the `DLRadioGroup` linked with radio buttons.
pr3

I changed the sample project to conform with this new approach.
Let me know what you think about it.

### CODE REVIEW

**Comment by DavydLiu** [comment]
Hi Felipe,

Thanks for taking the effort and putting up this pull request, really appreciate it :D. I'm gonna review it sometime in the weekend. Meanwhile, could you also be more specific regarding the downsides of selecting radio buttons programmatically?

**Comment by felipowsky** [comment]
The main reason for my implementation of `DLRadioGroup` was my need to select a radio button with a specific value (an enum).

To accomplish that using the current structure I create a `IBOutlet` for each one of my `DLRadioButton`. Then to really select of them I test my enum value and wrote a switch-case statement to select the right `DLRadioButton`.

At this point I thought it would be great to have a easier way to select the radio button so I start to develop `selectRadioButtonWithTag:` in `DLRadioButton`.
But implementing the selection of another button in other button sounds really strange, doesn't it?
Because if for some reason I need to have another `IBOutlet` to a `DLRadioButton` I need to check which one of them has the list of all `radioButtons` correctly configured or link all radio buttons of the group to all `DLRadioButton` (which is really boring and creates a huge possibility for mistakes).

Then I realized that the responsibility of selecting a radio button doesn't belong to a radio button but a "manager" which I called `DLRadioGroup`.
Actually I think there is a even better way to select a button by a custom value, probably using delegate. But for now `selectRadioButtonWithTag:` solved my problem.

Well that was my way of thinking and I hope it answers your question.

Let me know if I misunderstood anything. :)

File: DLRadioButton/DLRadioButton.h [hunk 1]
```diff
@@ -62,24 +60,10 @@ IB_DESIGNABLE
...
-/**
- @return Selected buttons in same group, use it only if multiple selection is enabled.
- */
-- (NSArray *)selectedButtons;
-
-/**
- Clears selection for other buttons in in same group.
-*/
-- (void)deselectOtherButtons;
+- (void)willAddToRadioGroup:(DLRadioGroup *)radioGroup;
```
Comment (DavydLiu) [inline DLRadioButton/DLRadioButton.h]: These method shouldn't be exposed to public interface.

File: DLRadioButton/DLRadioButton.m [hunk 2]
```diff
@@ -109,66 +87,28 @@ - (UIImage *)drawIconWithSelection:(BOOL)selected {
     return image;
 }
 
-- (void)touchDown {
-    [self setSelected:YES];
+- (void)touchDown
+{
+    NSAssert(self.radioGroup != nil, @"Radio button has to be in a radio group");
```
Comment (DavydLiu) [inline DLRadioButton/DLRadioButton.m]: Nice assertion.

File: DLRadioButton/DLRadioButton.m [hunk 3]
```diff
@@ -109,66 +87,28 @@ - (UIImage *)drawIconWithSelection:(BOOL)selected {
...
+    
+    [self.radioGroup selectRadioButton:self];
 }
 
-- (void)initRadioButton {
+- (void)initRadioButton
+{
```
Comment (DavydLiu) [inline DLRadioButton/DLRadioButton.m]: I need to move braces into next line in next iteration.

**Comment by DavydLiu** [comment]
Hi Felipe,

I really appreciate your enthusiasm; the approach is to pull things apart is very insightful! After review the code, I decided to not merge your pull request since changes to APIs are way too large. Consider over 400 apps are using `DLRadioButton` right now, I’m sure it will break a ton of them. 

The other reason would be I think one thing people love about `DLRadioButton` is its nimble size and straightforward interfaces. I’m pretty sure people don’t want copy four files and import two headers just to use a radio button :) So, it would be the best interest of every user to keep things simple and small.

Regarding your concern about “I need to check which one of them has the list of all radioButtons correctly configured”. I’m not sure that I understand it here, you can check any button in the same group to get the list of all radio buttons.

Lastly, I think you did a really good job on implementing `DLRadioGroup`. Just wondering if you’re willing to fork `DLRadioButton` with your radio group. I’d be more than happy to add your fork in README, so that if people do need a separate radio group can still use yours. I'll incorporate some of your ideas into `DLRaidoButton` in next iteration.

**Comment by felipowsky** [comment]
Thanks for replying and commenting @DavydLiu.
Indeed it's a big change on the API. I understand your concern.

> Regarding your concern about “I need to check which one of them has the list of all radioButtons correctly configured”. I’m not sure that I understand it here, you can check any button in the same group to get the list of all radio buttons.

Yes. But think about 2 radio buttons in the same group. If I understand it right, now you need to link all `radioButtons` for both of them if you want to have radio buttons correctly linked with their other buttons, right? I think this is a little annoying although I understand the purpose of keeping the API simple.

Feel free to mention my fork. :)
Maybe later we can think about a better way to smoothly migrate to my approach or have it available as an optional feature.

Again, thank you for your feedback.

**Comment by DavydLiu** [comment]
I don't think it is necessary to make sure **both** of them correctly linked with their other buttons. As long as you put the button in one of the other buttons. It will automatically chain them together. Also, I think it would be nice to have a method as much `addOtherButton:`, which will a button to any button's group.

**Pull request closed without merging**